### PR TITLE
TEL-5120: cleanup socket when disconnecting/close

### PR DESCRIPTION
**Pull request by StefanYohansson**

### CODE REVIEW

File: src/switch_rtp.c [hunk 1]
```diff
@@ -3264,7 +3263,10 @@ static void ping_socket(switch_rtp_t *rtp_session)
 {
 	uint32_t o = UINT_MAX;
 	switch_size_t len = sizeof(o);
-	switch_socket_sendto(rtp_session->sock_input, rtp_session->local_addr, 0, (void *) &o, &len);
+
+	if (rtp_session->sock_input) {
```
Comment (dev-ryanc) [inline src/switch_rtp.c:3267]: Please check as well the rtp_session

File: src/switch_rtp.c [hunk 2]
```diff
@@ -5528,25 +5530,38 @@ SWITCH_DECLARE(void) switch_rtp_break(switch_rtp_t *rtp_session)
...
+	switch_socket_t *sock_input, *sock_output;
 	switch_assert(rtp_session != NULL);
 	switch_mutex_lock(rtp_session->flag_mutex);
 	if (rtp_session->flags[SWITCH_RTP_FLAG_IO]) {
 		rtp_session->flags[SWITCH_RTP_FLAG_IO] = 0;
 		if (rtp_session->sock_input) {
 			ping_socket(rtp_session);
-			switch_socket_shutdown(rtp_session->sock_input, SWITCH_SHUTDOWN_READWRITE);
+			sock_input = rtp_session->sock_input;
+			switch_socket_shutdown(sock_input, SWITCH_SHUTDOWN_READWRITE);
```
Comment (dev-ryanc) [inline src/switch_rtp.c:5541]: There is a scenario where sock_output points to the same address as sock_input. Shutting it down here will make sock_output points to garbage memory. You may need to check sock_output before destroying sock_input

File: src/switch_rtp.c [hunk 3]
```diff
@@ -5528,25 +5530,38 @@ SWITCH_DECLARE(void) switch_rtp_break(switch_rtp_t *rtp_session)
...
+			switch_socket_close(sock_output);
 		}
 
 		if (rtp_session->flags[SWITCH_RTP_FLAG_ENABLE_RTCP]) {
-			if (rtp_session->rtcp_sock_input && rtp_session->rtcp_sock_input != rtp_session->sock_input) {
+			if (rtp_session->rtcp_sock_input && rtp_session->rtcp_sock_input != sock_input) {
 				ping_socket(rtp_session);
-				switch_socket_shutdown(rtp_session->rtcp_sock_input, SWITCH_SHUTDOWN_READWRITE);
+				sock_input = rtp_session->rtcp_sock_input;
+				switch_socket_shutdown(sock_input, SWITCH_SHUTDOWN_READWRITE);
```
Comment (dev-ryanc) [inline src/switch_rtp.c:5556]: There is a scenario where sock_output points to the same address as sock_input. Shutting it down here will make sock_output points to garbage memory. You may need to check sock_output before destroying sock_input

**Review by dev-ryanc — CHANGES_REQUESTED** [review]

**Review by piotrgregor — COMMENTED** [review]
The crash was at `zrtp_send_rtp_callback`. 
Are you not trying to fix the outcome instead of a root cause @StefanYohansson ? Isn't the root problem that zrtp callback gets called when socket is already closed (or in a malformed state, e.g. not started yet - notice it was called from `_send_and_resend_hello`)?
Maybe we should find out how to disable zrtp callback instead of fixing all the problems that will arise when callback hasn't been stopped properly?

**Pull request closed without merging**